util/rowcodec/decoder.go

```
@@ -183,31 +298,163 @@ func (decoder *Decoder) decodeColData(colIdx int, colData []byte, chk *chunk.Chu
 		j.Value = colData[1:]
 		chk.AppendJSON(colIdx, j)
 	default:
-		return errors.Errorf("unknown type %d", ft.Tp)
+		return errors.Errorf("unknown type %d", col.Tp)
 	}
 	return nil
 }
+
+// DecodeToBytes decodes raw byte slice to row data.
+func (decoder *Decoder) DecodeToBytes(outputOffset map[int64]int, handle int64, value []byte, cacheBytes []byte) ([][]byte, error) {
+	var r row
+	err := r.fromBytes(value)
+	if err != nil {
+		return nil, err
+	}
+	values := make([][]byte, len(outputOffset))
+	for _, col := range decoder.columns {
+		var err error
+		tp, err := fieldType2Flag(byte(col.Tp), uint(col.Flag))
+		if err != nil {
+			return nil, err
+		}
+		colID := col.ID
+		offset := outputOffset[colID]
+		if col.IsPKHandle || colID == model.ExtraHandleID {
+			handleData := cacheBytes
+			if decoder.forceVarint {
+				if mysql.HasUnsignedFlag(uint(col.Flag)) {
+					handleData = append(handleData, VaruintFlag)
+					handleData = codec.EncodeUvarint(handleData, uint64(handle))
+				} else {
+					handleData = append(handleData, VarintFlag)
+					handleData = codec.EncodeVarint(handleData, handle)
+				}
+			} else {
+				if mysql.HasUnsignedFlag(uint(col.Flag)) {
+					handleData = append(handleData, UintFlag)
+					handleData = codec.EncodeUint(handleData, uint64(handle))
+				} else {
+					handleData = append(handleData, IntFlag)
+					handleData = codec.EncodeInt(handleData, handle)
+				}
+			}
+			values[offset] = handleData
+			continue
+		}
+
+		idx, isNil, notFound := r.findColID(colID)
+		if !notFound && !isNil {
+			val := r.getData(idx)
+			values[offset] = decoder.encodeOldDatum(tp, val)
+			continue
+		}
+
+		if isNil {
+			values[offset] = []byte{NilFlag}
+			continue
+		}
+
+		if col.DefaultValue != nil {
+			defVal, err := col.DefaultValue()
+			if err != nil {
+				return nil, err
+			}
+			if len(defVal) > 0 {
+				values[offset] = decoder.encodeOldDatum(tp, defVal)
+				continue
+			}
+		}
+		if mysql.HasNotNullFlag(uint(col.Flag)) {
```

If the column not found and it doesn't have a default value, should we return an error directly? Why should we check the NotNullFlag?